Source/Libraries/Adapters/GrafanaAdapters/Functions/BuiltIn/AddMetaData.cs

```diff
@@ -0,0 +1,77 @@
+﻿using GrafanaAdapters.DataSourceValueTypes;
+using GrafanaAdapters.DataSourceValueTypes.BuiltIn;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace GrafanaAdapters.Functions.BuiltIn;
+
+/// <summary>
+/// Returns a series with an extra MetaData Field.
+/// </summary>
+/// <remarks>
+/// Signature: <c>AddMetaData(field, value, expression)</c><br/>
+/// Returns: Series of values.<br/>
+/// Example: <c>AddMetaData('Company', 'GPA', FILTER ActiveMeasurements WHERE SignalType='FREQ')</c><br/>
+/// Execution: Deferred enumeration.
+/// </remarks>
+public abstract class AddMetaData<T> : GrafanaFunctionBase<T> where T : struct, IDataSourceValueType<T>
+{
+    /// <inheritdoc />
+    public override string Name => nameof(AddMetaData<T>);
+
+    /// <inheritdoc />
+    public override string Description => "Returns a series of values with an extra metadata field.";
+
+    /// <inheritdoc />
+    public override ReturnType ReturnType => ReturnType.Series;
+
+    /// <inheritdoc />
+    public override ParameterDefinitions ParameterDefinitions => new List<IParameter>
+    {
+        new ParameterDefinition<string>
+        {
+            Name = "field",
+            Default = "Field",
+            Description = "A string containing the name of the field to be added.",
+            Required = true
+        },
+        new ParameterDefinition<string>
+        {
+            Name = "value",
+            Default = "Value",
+            Description = "A string representing the value of the field to be added.",
+            Required = true
+        },
+    };
+
+    /// <inheritdoc />
+    public override async IAsyncEnumerable<T> ComputeAsync(Parameters parameters, CancellationToken cancellationToken)
+    {
+        Dictionary<string, MetadataMap> metadataMaps = parameters.MetadataMaps;
+        string field = parameters.Value<string>(0);
+        string value = parameters.Value<string>(1);
+
+        await using IAsyncEnumerator<T> enumerator = GetDataSourceValues(parameters).GetAsyncEnumerator(cancellationToken);
+
+        while (await enumerator.MoveNextAsync().ConfigureAwait(false))
+        {
+            if (metadataMaps.TryGetValue(enumerator.Current.Target, out MetadataMap currentCoordinates))
```

Comment:
There are no operational restrictions on this function which means it can operate on a slice, set or single series. What that means in terms of metadata targeted by the function is that metadata being processed by the computation will vary based on the selected group operation. For example, there will be a "single" set of metadata for the common single series operation, vs., metadata for _all series_ being available in the slice and set operations. 

Regardless, repeating assignment of the same metadata over and over for each series is a bit CPU wasteful since the metadata is a singular set per data source value target for a given query operation. 

If this is truly an `AddMetadata` operation - and not an "augment - as the function name and description imply, code should simply add when metadata when it has been determined that it does not exist, e.g.:

```c#
if (metadataMaps.TryGetValue(enumerator.Current.Target, out MetadataMap currentCoordinates) && !currentCoordinates.ContainsKey(field))
    currentCoordinates.Add(field, value);
```